### PR TITLE
ClosestPoint to Linestring Updates for Reverse Searching with Cut Off

### PR DESCRIPTION
**Pull request by kevinkreiser**
I was using this function in another project and it came to a point where i wanted to consider not just some distance ahead of the `begin_shape_index` but also some distance behind it. So I've updated the function to allow for searching some distance in reverse. The defaults to the function insure that it works as it did previously (only search in the forward direction and assume you have no distance cut off).

If you specify a `reverse_dist_cutoff` that is `> 0` then the updated code will do a second pass iterating backwards from the shape index specified. i took special care to make sure that the performance of the method wouldn't be impacted in that the defaults should make it so that only a few additional conditionals will be checked before returning.

I also did an optimization where I used the cached result of the longitude scale from the distance approximator instead of doing that `std::cosf` twice. So we save a trig function call now.

One other small thing I did here was a bit of formatting to the unit tests to actually print the name of the test on the failure and some formatting so that the brackets look the same as the rest of the build steps.

In terms of testing what I did was just take the closest point tests and run them again but in the reverse direction. It was actually quite tricky with respect to edge cases to get the expected indices right. In fact I ended up changing the logic for who wins on a tie so that between forward and reverse searching the results were more intuitive. (This is so hard to describe in text).

## Tasklist

 - [x] add tests
 - [ ] review - you must request approval to merge any PR to master
 - [ ] ~Add #fixes with the issue number that this PR addresses~
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] ~Add to release notes in valhalla-docs~
 - [ ] ~update relevant documentation if this is an API impacting change~

### CODE REVIEW

File: src/midgard/pointll.cc [hunk 1]
```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
...
 
-  // If there is one point to process we are done
-  if (process_size == 1) {
-    return std::make_tuple(pts[begin_index], std::sqrt(DistanceSquared(pts[begin_index])),
-                           begin_index);
-  }
-
-  // Longitude (x) is scaled by the cos of the latitude so that distances are
-  // correct in lat,lon space
-  float lon_scale = cosf(lat() * kRadPerDeg);
```
Comment (kevinkreiser) [inline src/midgard/pointll.cc]: feels good to remove this trig function

File: src/midgard/pointll.cc [hunk 2]
```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
...
-      point = {v.lng(), v.lat()};
-    } // Projects along the ray between u and v
-    else {
-      scale /= sq;
-      point = {u.lng() + bx * scale, u.lat() + by * scale};
-    }
+  float mindistsqr = approx.DistanceSquared(closest);
+
+  // start going backwards, then go forwards
+  for (int reverse = 0; reverse < 2; ++reverse) {
```
Comment (kevinkreiser) [inline src/midgard/pointll.cc]: i could make tuples out of the 3 things below and iterate over a vector of 2 tuples. this might be more performant but is probably in the noise

File: src/midgard/pointll.cc [hunk 3]
```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
...
-    // Check if this point is better
-    const auto sq_distance = approx.DistanceSquared(point);
-    if (sq_distance < mindistsqr) {
-      closest_segment = static_cast<int>(index);
-      mindistsqr = sq_distance;
-      closest = std::move(point);
-    }
+      // Check if this point is better
+      const auto sq_distance = approx.DistanceSquared(point);
+      if (sq_distance <= mindistsqr) {
```
Comment (kevinkreiser) [inline src/midgard/pointll.cc]: `<=` is the tie braker. it helps when trying to think about what index should be returned when the test point _is_ a point in the linestring

File: src/midgard/pointll.cc [hunk 4]
```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
...
+        closest_segment = index - reverse;
```
Comment (kevinkreiser) [inline src/midgard/pointll.cc]: `- reverse` is so that you get the left most (floor index) for the segment you are on even when searching in reverse

File: src/midgard/pointll.cc [hunk 5]
```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
...
+        mindistsqr = sq_distance;
+        closest = std::move(point);
+      }
 
-    // Check if we should bail early because of looking at too much shape
-    if (dist_cutoff != std::numeric_limits<float>::infinity() && (dist_cutoff -= u.Distance(v)) < 0) {
-      break;
```
Comment (kevinkreiser) [inline src/midgard/pointll.cc]: we just check the size of dist_cutoff in the loop now, its cleaner this way

File: src/midgard/pointll.cc (hunk 2 above)
Comment (danpat) [reply src/midgard/pointll.cc]: I think the ternary expressions below are OK, they're pretty self contained.

*But*, I think it would be a bit more readable to use this form of the loop and make `reverse` a boolean:

`for (bool reverse : { true, false })`

Comment (dnesbitt61) [reply src/midgard/pointll.cc]: haven't seen this before but I like it
